iroh-sync/src/store/fs.rs

```diff
@@ -237,6 +250,72 @@ impl super::Store for Store {
     fn content_hashes(&self) -> Result<Self::ContentHashesIter<'_>> {
         ContentHashesIterator::create(&self.db)
     }
+
+    fn register_useful_peer(&self, namespace: NamespaceId, peer: crate::PeerIdBytes) -> Result<()> {
+        // calculate nanos since UNIX_EPOCH for a time measurement
+        let nanos = std::time::UNIX_EPOCH
+            .elapsed()
+            .map(|duration| duration.as_nanos() as u64)?;
+        let write_tx = self.db.begin_write()?;
+        {
+            let mut peers_table = write_tx.open_multimap_table(NAMESPACE_PEERS_TABLE)?;
+            let namespace_peers = peers_table.get(namespace.as_bytes())?;
+
+            // find any previous entry for the same peer to remove it
+            let mut prev_peer_nanos = None;
+            // calculate the len in the same loop since calling `len` is another fallible operation
+            let mut len = 0;
+            // the oldest entry in the table, candidate for removal depending on len
+            let mut oldest_entry: Option<(Nanos, PeerIdBytes)> = None;
+            for result in namespace_peers {
```

Comment:
I don't think you need to iterate over all values, I think you can just take the first one. Per the docs of https://docs.rs/redb/latest/redb/trait.ReadableMultimapTable.html#tymethod.get the iterator is sorted in ascending order, and the order for tuples is that they are element wise. So if the timestamp is the first element in the value, the oldest value should be the first one of the iterator.

Reply:
Oh nice I mistook this for descending. I still need to iterate the values to find a previous entry for the same peer though. But candidate for removal can make use of this fact, ty!

Reply:
Also MultimapValue implments https://doc.rust-lang.org/nightly/core/iter/trait.DoubleEndedIterator.html so it is cheap to get the last element as well

Reply:
done in [d84c374](https://github.com/n0-computer/iroh/pull/1564/commits/d84c374814628da767f458d1b84062e9a059800a)